### PR TITLE
Add `log_scale` option of Normal Distritbuion

### PR DESCRIPTION
**Pull request by YoshikawaMasashi**
I added `ln_var` option to `chainer.distributions.Normal`.

### CODE REVIEW

File: chainer/distributions/normal.py [hunk 1]
```diff
@@ -32,14 +32,27 @@ class Normal(distribution.Distribution):
...
+            raise ValueError(
+                "Either `scale` or `ln_var` (not both) must have a value.")
         self.loc = chainer.as_variable(loc)
-        self.scale = chainer.as_variable(scale)
+        if scale is None:
+            self.ln_var = chainer.as_variable(ln_var)
+            self.scale = exponential.exp(0.5 * self.ln_var)
+        else:
+            self.scale = chainer.as_variable(scale)
+            self.ln_var = 2. * exponential.log(self.scale)
```
Comment (toslunar) [inline chainer/distributions/normal.py]: Could you compute the cache (`scale` or `ln_var`) with config `enable_backprop`?  Or memorize the config and check if it's kept when methods are called.

**Comment by toslunar** [comment]
On the name of the argument, `log_var` looks better since `Distribution` has `log_prob` method, which would precede the fact `F.gaussian` has `ln_var`.  What do you think?

**Comment by YoshikawaMasashi** [comment]
Um, I think that what you are saying is reasonable. I'll fix.

**Comment by YoshikawaMasashi** [comment]
jenkins, test this please.

File: tests/chainer_tests/distributions_tests/test_kldivergence.py [hunk 2]
```diff
@@ -74,5 +82,16 @@ def test_normal_normal_gpu(self):
         dist2 = self.make_normal_dist(True)
         self.check_kl(dist1, dist2)
 
+    def test_normal_normal_log_var_cpu(self):
+        dist1 = self.make_normal_dist_log_var()
+        dist2 = self.make_normal_dist_log_var()
+        self.check_kl(dist1, dist2)
+
+    @attr.gpu
+    def test_normal_normal__log_var_gpu(self):
```
Comment (toslunar) [inline tests/chainer_tests/distributions_tests/test_kldivergence.py]: Fix typo (`__` -> `_`).

File: chainer/distributions/normal.py [hunk 3]
```diff
@@ -32,14 +32,29 @@ class Normal(distribution.Distribution):
...
+        have a value.
+        log_var(:class:`~chainer.Variable` or :class:`numpy.ndarray` or \
+        :class:`cupy.ndarray`): Parameter of distribution representing the \
+        scale :math:`\\log(\\sigma^2)`. Either `scale` or `log_var` (not \
+        both) must have a value.
 
     """
 
-    def __init__(self, loc, scale):
+    def __init__(self, loc, scale=None, log_var=None):
```
Comment (toslunar) [inline chainer/distributions/normal.py]: Would you like to make `log_var` keyword-only?  (If so, use `chainer.utils.argument.parse_kwargs`).

File: chainer/distributions/normal.py [hunk 4]
```diff
@@ -32,14 +32,29 @@ class Normal(distribution.Distribution):
...
         self.loc = chainer.as_variable(loc)
-        self.scale = chainer.as_variable(scale)
+
+        with chainer.using_config('enable_backprop', True):
+            if scale is None:
+                self.log_var = chainer.as_variable(log_var)
+                self.scale = exponential.exp(0.5 * self.log_var)
+            else:
+                self.scale = chainer.as_variable(scale)
+                self.log_var = 2. * exponential.log(self.scale)
```
Comment (toslunar) [inline chainer/distributions/normal.py]: Could you make `scale` and `log_var` read-only, or define setters of them to do this always?

BTW, it seems not bad to make all attributes of `Distribution` subclasses read-only.

File: tests/chainer_tests/distributions_tests/test_kldivergence.py [hunk 5]
```diff
@@ -74,5 +82,16 @@ def test_normal_normal_gpu(self):
...
+        dist1 = self.make_normal_dist_log_var(True)
+        dist2 = self.make_normal_dist_log_var(True)
+        self.check_kl(dist1, dist2)
+
```
Comment (toslunar) [inline tests/chainer_tests/distributions_tests/test_kldivergence.py]: Could you add KL tests with mixed argument sets?
To do so, it'll be easier if `make_normal_dist_log_var` is `make_normal_dist` with a new argument (not a new method).

File: tests/chainer_tests/distributions_tests/test_normal.py [hunk 6]
```diff
@@ -27,8 +28,12 @@ def setUp_configure(self):
         loc = numpy.random.uniform(-1, 1, self.shape).astype(numpy.float32)
         scale = numpy.exp(
             numpy.random.uniform(-1, 1, self.shape)).astype(numpy.float32)
-        self.params = {"loc": loc, "scale": scale}
-        self.scipy_params = {"loc": loc, "scale": scale}
+        if self.log_var_option:
+            self.params = {"loc": loc, "log_var": 2. * numpy.log(scale)}
+            self.scipy_params = {"loc": loc, "scale": scale}
```
Comment (toslunar) [inline tests/chainer_tests/distributions_tests/test_normal.py]: `self.scipy_params` doesn't depend on `log_var_option`.

File: chainer/distributions/normal.py [hunk 7]
```diff
@@ -122,6 +137,6 @@ def variance(self):
 
 @distribution.register_kl(Normal, Normal)
 def _kl_normal_normal(dist1, dist2):
-    return exponential.log(dist2.scale) - exponential.log(dist1.scale) \
+    return 0.5 * dist2.log_var - 0.5 * dist1.log_var \
```
Comment (toslunar) [inline chainer/distributions/normal.py]: Ditto.

File: tests/chainer_tests/distributions_tests/test_kldivergence.py [hunk 8]
```diff
@@ -67,31 +65,26 @@ def test_laplace_laplace_cpu(self):
...
-        dist1 = self.make_laplace_dist(True)
-        dist2 = self.make_laplace_dist(True)
-        self.check_kl(dist1, dist2)
+        for use_log_scale1 in [True, False]:
+            for use_log_scale2 in [True, False]:
+                dist1 = self.make_normal_dist(
+                    True, use_log_scale=use_log_scale1)
+                dist2 = self.make_normal_dist(
+                    True, use_log_scale=use_log_scale2)
+                self.check_kl(dist1, dist2)
```
[the remark this answers is not in the document]
Comment (YoshikawaMasashi) [reply tests/chainer_tests/distributions_tests/test_kldivergence.py]: Aaahh, fixed.

**Comment by toslunar** [comment]
Jenkins, test this please.

**Pull request merged**